src/services/asset-discovery-service.ts

```diff
@@ -79,8 +105,10 @@ export default class AssetDiscoveryService extends PercyClientService {
     profile('--> assetDiscoveryService.discoverForWidths', {url: rootResourceUrl})
     const resourcePromises: Array<Promise<any[]>> = []
     for (let idx = 0; idx < widths.length; idx++) {
+      const page = this.pages !== null ? this.pages[idx] : null
```

Comment:
I think this could be problematic.

Let's say we have two snapshot requests that come in. Each with 7 widths. We'll assume we have `isParallel` set to `true`.

* Before any snapshots come in, we've created 10 pages.
* The first snapshot request comes in and the first 7 pages are allocated
* While that begins, a second snapshot request comes in. It looks like we'll reallocate those same 7 pages even though they might be in use?

Reply:
A resource pool would definitely resolve these issues.

What you would do is specify the upper bounds of your pool... I actually think we could make it just 5 pages by default because the vast majority of folks take snapshots at 2-3 widths only (and no parallel requests). We could potentially make the max pool size configurable with a command line flag so people could control the max resources we'll burst up to.

And then if the pool is full, we'll have to wait until a page becomes free to use. And that mechanism will be how you can handle parallel snapshot requests. Worst case, you wait a little bit.

I did a bit of Googling around. https://github.com/coopernurse/node-pool was the first thing that came up. There might be other options out there too.

Would be curious to get your thoughts on all this @blake-newman

Reply:
I believe you are right, resolving this way would mean that there could be a clash, thus this is the exact problem with parallel tests currently. Albeit this logic is now only available in serial execution, thus making hitting the same page in the pool near (probably) impossible. 

The main issue trying to be solved is ensuring that there is no clash with requests on the same page, which with a pool it likely to be observed. IMO I believe this solution does justice to enabling parallel builds with the minimal amount of changes, to prevent likely bugs. However, I don't believe it is the best solution, as a revolving pool of pages will ensure optimised run times for parallel builds. 

By the nature of parallel builds running a magnitude faster, i believe we should aim to support parallel builds at a minimum. For example we have a 5 mins test suite, enabling parallel builds on our CI will bring the results down to around 1-2mins which is a huge win in a CI environment. In this case the changes here would need to introduce roughly 3 mins of time. Based on 75ms load time for a page this would mean 2400 snapshots in the whole test run (:money_mouth_face:), thus i don't think it's a huge problem to load on demand. 

Obviously, i'm happy to work on a full solution that both suits sequential and parallel. I'm a big fan of delivering smaller change sets to enable value, i would be concerned that introducing a more complex solution to enable parallel builds will cause unexpected bugs that we can't forsee. However, this is your codebase so happy to do either to get this in to improve the lives for every engineer. 

Tackling the points raised, i believe the `node-pool` which i also looked into would be a great solution. Although i think it would require some large code changes to ensure we cope with scheduling.

Reply:
@djones i have pushed changes to use `generic-pool` as i found it pretty simple to implement. 

The question now is how big we make the pool. Memory issues we was facing was not a problem with percy but due to this issue: https://github.com/cypress-io/cypress/issues/350

So i believe setting to `min: 2, and max: 10` should be sufficient. However a larger pool will probably improve performance with this solution as there will be more pages to use at any given time. So for parallel builds this will be more optimised. Perhaps giving the option to set max pool size via `PERCY_POOL_SIZE` will allow configuration for low / high memory environments.